lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -136,15 +146,246 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
+  session_state.set_current_version(++cur_version, update_criteria);
+  MLOG(MINFO) << "State of session changed to "
+               << session_fsm_state_to_str(session_state.get_state())
+	       << " Current version is " << session_state.get_current_version();
+
   /* Update the m5gsm_cause and prepare for respone along with actual cause*/
   prepare_response_to_access(
       imsi, session_state, magma::lte::M5GSMCause::OPERATION_SUCCESS);
-  pipelined_client_->set_upf_session(sess_info, call_back_void_upf);
+
+  SessionState::SessionInfo sess_info;
+  sess_info.imsi       = imsi;
+  sess_info.ip_addr    = ip_addr;
+  sess_info.Pdr_rules_ = session_state.get_all_pdr_rules();
+  sess_info.Far_rules_ = session_state.get_all_far_rules();
+  session_state.sess_infocopy(&sess_info);
+  pipelined_client_->set_upf_session(sess_info, call_back_upf);
   return;
 }
 
+/* Function to initiate release of the session in enforcer requested by AMF
+ * Go over session map and find the respective session of imsi
+ * Go over SessionState vector and find the respective dnn (apn) 
+ * start terminating session process
+ */
+bool SessionStateEnforcer::m5g_release_session(
+      SessionMap& session_map, const std::string& imsi,
+      const std::string& dnn, SessionUpdate& session_update)
+{
+  /* Search with session search criteria of IMSI and apn/dnn and 
+   * find  respective sesion to release operation
+   */
+  SessionSearchCriteria criteria(imsi, IMSI_AND_APN, dnn);
+  auto session_it = session_store_.find_session(session_map, criteria);
+  if(!session_it) {
+     MLOG(MERROR) << "No session found in SessionMap for IMSI " << imsi
+                  << " with DNN " << dnn << " to release"; 
+     return false;
+  } else {
+     //Found the respective session to be updated
+     auto& session   = **session_it;
+     auto session_id = session->get_session_id();
+     /*Irrespective of any State of Session, release and terminate*/
+     SessionStateUpdateCriteria& session_uc = session_update[imsi][session_id];
+     MLOG(MINFO)<< "Trying to release session with id " << session_id
+		<< " from state " 
+		<< session_fsm_state_to_str(session->get_state());
+     m5g_start_session_termination(imsi, session, dnn, session_uc);
+  }
+  return true;
+}
+
+/*Start processing to terminate respective session requested from AMF*/
+void SessionStateEnforcer::m5g_start_session_termination(
+      const std::string& imsi, const std::unique_ptr<SessionState>& session,
+      const std::string& dnn, SessionStateUpdateCriteria& session_uc)
+{
+   auto session_id = session->get_session_id();
+
+  /* update respective session's state and return from here before timeout
+   * to update session store with state and version
+   */
+  session->set_fsm_state(RELEASE, session_uc);
+  uint32_t cur_version = session->get_current_version();
+  session->set_current_version(cur_version++, session_uc);
+  MLOG(MINFO) << "During release state of session changed to "

```

Comment:
This log is good, but might lead to duplicate logging because `set_fsm_state` also logs the state change. Maybe it would be good to log the version change in `set_current_version` so we can get rid of this log below.

Reply:
changed.